launchpad/src/launchpad.rs

```diff
@@ -1,9 +1,517 @@
+                    self.blockchain().get_prev_block_random_seed(),
+                    self.blockchain().get_block_random_seed(),
+                ),
+                VEC_MAPPER_START_INDEX,
+            ),
+            OngoingOperationType::SelectWinners {
+                seed,
+                seed_index,
+                ticket_position,
+            } => {
+                self.clear_operation();
+
+                (
+                    Random::from_hash(self.crypto(), seed, seed_index),
+                    ticket_position,
+                )
+            }
+            _ => return sc_error!("Another ongoing operation is in progress"),
+        };
+
+        let last_ticket_position = self.shuffled_tickets().len();
+        let nr_winning_tickets = self.nr_winning_tickets().get();
+
+        let gas_before = self.blockchain().get_gas_left();
+
+        let is_winning_ticket = ticket_position <= nr_winning_tickets;
+        self.shuffle_single_ticket(
+            &mut rng,
+            ticket_position,
+            last_ticket_position,
+            is_winning_ticket,
+        );
+        ticket_position += 1;
+
+        let gas_after = self.blockchain().get_gas_left();
+        let gas_per_iteration = gas_before - gas_after;
+
+        while ticket_position < last_ticket_position - 1 {
+            if self.can_continue_operation(gas_per_iteration) {
+                let is_winning_ticket = ticket_position <= nr_winning_tickets;
+                self.shuffle_single_ticket(
+                    &mut rng,
+                    ticket_position,
+                    last_ticket_position,
+                    is_winning_ticket,
+                );
+                ticket_position += 1;
+            } else {
+                self.save_progress(&OngoingOperationType::SelectWinners {
+                    seed: rng.seed,
+                    seed_index: rng.index,
+                    ticket_position,
+                });
+
+                return Ok(OperationCompletionStatus::InterruptedBeforeOutOfGas);
+            }
+        }
+
+        self.start_confirmation_period(VEC_MAPPER_START_INDEX, nr_winning_tickets);
+
+        Ok(OperationCompletionStatus::Completed)
+    }
+
+    #[payable("*")]
+    #[endpoint(confirmTickets)]
+    fn confirm_tickets(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_amount] payment_amount: Self::BigUint,
+        nr_tickets_to_confirm: usize,
+    ) -> SCResult<OperationCompletionStatus> {
+        let current_epoch = self.blockchain().get_block_epoch();
+        let claim_start_epoch = self.claim_period_start_epoch().get();
+        let claim_period_in_epochs = self.claim_period_in_epochs().get();
+        let claim_end_epoch = claim_start_epoch + claim_period_in_epochs;
+
+        if current_epoch > claim_end_epoch {
+            self.select_new_winners()?;
+            return Ok(OperationCompletionStatus::InterruptedBeforeOutOfGas);
```

Comment:
I do not thing this message is correct always. select_new_winners could return simply error if confirming was problematic. You need to return error here in order to return the payment automatically.